packages/components/src/components/as-category-widget/as-category-widget.spec.ts

```diff
@@ -0,0 +1,185 @@
+import { TestWindow } from '@stencil/core/dist/testing';
+import { Category, CategoryWidget } from './as-category-widget';
+
+describe('as-category-widget', () => {
+  let categoryWidget;
+
+  beforeEach(() => {
+    categoryWidget = new CategoryWidget();
+  });
+
+  it('should build', () => {
+    expect(new CategoryWidget()).toBeTruthy();
+  });
+
+  describe('Rendering', () => {
+    let element: HTMLAsCategoryWidgetElement;
+    let testWindow: TestWindow;
+
+    beforeEach(async () => {
+      testWindow = new TestWindow();
+      element = await testWindow.load({
+        components: [CategoryWidget],
+        html: '<as-category-widget></as-category-widget>'
+      });
+    });
+
+    it('should render properly', async () => {
+      element.heading = 'Category Widget Example';
+      element.description = 'Description for Category Widget';
+      element.categories = exampleCategories;
+      element.showClearButton = true;
+      await testWindow.flush();
+
+      expect(element).toMatchSnapshot();
+    });
+
+    it('should not render header when showHeader is false', async () => {
+      element.heading = 'Category Widget Example';
+      element.description = 'Description for Category Widget';
+      element.showHeader = false;
+      await testWindow.flush();
+
+      expect(element).toMatchSnapshot();
+    });
+
+    it('should render clear button', async () => {
+      element.showClearButton = true;
+      element.categories = exampleCategories;
+      await testWindow.flush();
+
+      expect(element).toMatchSnapshot();
+    });
+
+    it('should render bars according to total value', async () => {
+      element.categories = exampleCategories;
+      element.useTotalPercentage = true;
+      await testWindow.flush();
+
+      expect(element).toMatchSnapshot();
+    });
+  });
+
+  describe('.getSelectedCategories', () => {
+    it('should return selectedCategories', () => {
+      const selectedCategories = ['Category 1'];
+      categoryWidget.selectedCategories = selectedCategories;
+
+      expect(categoryWidget.getSelectedCategories()).toEqual(selectedCategories);
+    });
+  });
+
+  describe('.clearSelection', () => {
+    beforeEach(() => {
+      spyOn(categoryWidget, '_onCategoriesChanged');
+    });
+
+    it('should clear selected categories', () => {
+      const selectedCategories = ['Category 1'];
+      categoryWidget.selectedCategories = selectedCategories;
+
+      categoryWidget.clearSelection();
+
+      expect(categoryWidget.getSelectedCategories()).toEqual([]);
+      expect(categoryWidget._onCategoriesChanged).toHaveBeenCalled();
```

Comment:
I think we want to test that the component fires some kind of event when the categories are cleared. I think with this expect, we are [testing the implementation](https://teamgaslight.com/blog/testing-behavior-vs-testing-implementation) but not the desired behavior.

I think we could create 2 new tests for .clearSelection() checking the component should trigger an event with previously selected categories and...with no selected categories should the component trigger an event? 🤔

Reply:
It should always trigger an event when categories are updated, to keep the user updated about the state of the widget.

I only need to modify the current test to listen to the fired event.

As I am testing a similar thing in `._onCategoriesChanged` describe section, I am going to rename that section to `Behaviour` and place that test there.

Reply:
No, that's not expected, I thought you meant other thing 🤦‍♂️